ios/browser/api/bookmarks/exporter/bookmark_html_writer.cc

```
@@ -0,0 +1,636 @@
+    if (!file_->IsValid()) {
+      PLOG(ERROR) << "Could not create " << path_;
+      return false;
+    }
+    return true;
+  }
+
+  // Increments the indent.
+  void IncrementIndent() {
+    indent_.resize(indent_.size() + kIndentSize, ' ');
+  }
+
+  // Decrements the indent.
+  void DecrementIndent() {
+    DCHECK(!indent_.empty());
+    indent_.resize(indent_.size() - kIndentSize, ' ');
+  }
+
+  // Called at the end of the export process.
+  void NotifyOnFinish(BookmarksExportObserver::Result result) {
+    if (observer_ != nullptr) {
+      observer_->OnExportFinished(result);
+    }
+  }
+
+  // Writes raw text out returning true on success. This does not escape
+  // the text in anyway.
+  bool Write(const std::string& text) {
+    if (!text.length())
+      return true;
+    size_t wrote = file_->WriteAtCurrentPos(text.c_str(), text.length());
+    bool result = (wrote == text.length());
+    if (!result) {
+      PLOG(ERROR) << "Could not write text to " << path_;
+      return false;
+    }
+    return true;
+  }
+
+  // Writes out the text string (as UTF8). The text is escaped based on
+  // type.
+  bool Write(const std::string& text, TextType type) {
+    DCHECK(base::IsStringUTF8(text));
+    std::string utf8_string;
+
+    switch (type) {
+      case ATTRIBUTE_VALUE:
+        // Convert " to &quot;
+        utf8_string = text;
+        base::ReplaceSubstringsAfterOffset(&utf8_string, 0, "\"", "&quot;");
+        break;
+
+      case CONTENT:
+        utf8_string = net::EscapeForHTML(text);
+        break;
+
+      default:
+        NOTREACHED();
+    }
+
+    return Write(utf8_string);
+  }
+
+  // Indents the current line.
+  bool WriteIndent() {
+    return Write(indent_);
+  }
+
+  // Converts a time string written to the JSON codec into a time_t string
+  // (used by bookmarks.html) and writes it.
+  bool WriteTime(const std::string& time_string) {
+    int64_t internal_value;
+    base::StringToInt64(time_string, &internal_value);
```

maybe throw an error if `StringToInt64` returns false? https://source.chromium.org/chromium/chromium/src/+/master:base/strings/string_number_conversions.h;l=60?q=StringToInt64&ss=chromium

---

It is from here: https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/bookmarks/bookmark_html_writer.cc;l=316;drc=929e790ad17a54008b3dac86cbe30751c968c2ce?q=html_writer&ss=chromium%2Fchromium%2Fsrc

The only thing changed in the `html_writer` is that I added code to write from a `unique_ptr<base::Value>` which is the `WriteBookmarks` function.